extensions/pkg/controller/worker/genericactuator/actuator_reconcile.go

```diff
@@ -348,16 +357,18 @@ func (a *genericActuator) waitUntilWantedMachineDeploymentsAvailable(ctx context
 			numDesired += existingMachineDeployment.Spec.Replicas
 			numUpdated += existingMachineDeployment.Status.UpdatedReplicas
 			numAvailable += existingMachineDeployment.Status.AvailableReplicas
+			numUnavailable += existingMachineDeployment.Status.UnavailableReplicas
 		}
 
 		var msg string
-
 		switch {
 		case !controller.IsHibernated(cluster):
-			if numUpdated >= numDesired && int(numHealthyDeployments) == len(wantedMachineDeployments) {
+			// numUpdated == numberOfAwakeMachines waits until the old machine is deleted in the case of a rolling update with maxUnavailability = 0
+			// numUnavailable == 0 makes sure that every machine joined the cluster (during creation & in the case of a rolling update with maxUnavailability > 0)
+			if numUnavailable == 0 && numUpdated == numberOfAwakeMachines && int(numHealthyDeployments) == len(wantedMachineDeployments) {
```

Comment:
There still could be one condition that hasn't been captured. The terminating machines (could be due to larger drain timeouts and PDBs) from the old machineSets. I think it might be okay to ignore if you think it's not too important as they are only terminating machines and new machines are already created. 

To capture something like that, we might need to check all machines belonging to the old machineSet and make sure there sum is 0.

Reply:
Hm, why was this changed? IIUC it's not related to the `RollingUpdate` condition. Is this something that we have to improve, i.e., was the previous check not good enough? cc @prashanth26 

```golang
			if numUpdated >= numDesired && int(numHealthyDeployments) == len(wantedMachineDeployments) {	
```

Reply:
> This changes the logic that waits for the rolling update to finish.
Currently when triggering a rolling update (e.g volume resize) with maxUnavailability = 0, it does not wait for the new machine to join the cluster. This is a bug in my opinion (see Release notes).

In certain cases it did not wait for the Rolling update to complete. So the RollingUpdate condition was removed too early.

Reply:
@rfranzke you are right, it is not strictly related to putting a condition into the Worker status. But there is little use to such a condition, if the check for a rolling update itself does not work as expected.

Reply:
Yes, I think having `numUnavailable == 0` should improve capturing this rolling update for longer.

Reply:
> There still could be one condition that hasn't been captured. The terminating machines (could be due to larger drain timeouts and PDBs) from the old machineSets. 

This is catched by `numUnavailable == 0`

Also added checks prevent issues with a stale cache.

Reply:
@prashanth26 can you please check this code again?

Reply:
I can also provide a list of the test cases I used to verify this PR if that helps @prashanth26 .

Reply:
Yes with `numUnavailable == 0` it looks better. Thanks. 

> I can also provide a list of the test cases I used to verify this PR if that helps @prashanth26 .

Do you this handy?

Reply:
- Worker Pool deletion
	- wait until unwanted machines have been deleted
- Worker Pool creation
	-  does not set rolling update condition to true

**Causing a rolling update (new MachineClass)** -> e.g Volume resize
- Max unavailability = 0 
	- detects rolling update, waits for machines to join, then waits for old machines to be deleted (observe in logs)
	- With PDB that does not allow deletion
		- Wait for deletion of machine (force delete)
	- with same pre-existing MachineClass (e.g due to failed reconciliation / crash)
		- because we are using the MachineClass as an indicator, this is not properly detected as a rolling update
		- should nontheless wait for the update to be ready
- Creation of worker in multiple avail. zones
	- increasing the worker count only scales up an already existing MachineSet. Should not be a rolling update
- Max Unavailability = 1
	- Wait for the new machine to join
- MCM is stuck simulation
	- simulate MachineDeployment that has no corresponding MachineSet created yet. Simulated with a MachineDeployment that the MCM ignores - which has no finalizer. Temporarely disabled the finalizer check in the extension provider.
	- should detect a stuck MCM and should restart the MCM pod.

Reply:
I think scenarios are exhaustive enough for now. looks good to me.

Reply:
I know it is quite some work, but I would be glad if someone would again test those (& possibly other scenarios that I did not think of). 
If not I ll do it again once all comments are resolved.

Reply:
Hi @danielfoehrKn ,

I can try to do it. However, I might be OOO until Tuesday. Would only be able to test the changes on Tuesday.

Reply:
I tested some scenarios, seems to be working well. Looks good to me.

Reply:
Thanks a lot for testing!